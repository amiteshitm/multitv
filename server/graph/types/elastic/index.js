import SoftwareES from './software'

export default SoftwareES
