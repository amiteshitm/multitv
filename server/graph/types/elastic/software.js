import elasticsearch from 'elasticsearch'
import { composeWithElastic } from 'graphql-compose-elasticsearch'

import mapping from 'shared/elastic/mapping'

require('dotenv').config()

const SoftwareElasticTC = composeWithElastic({
  graphqlTypeName: 'SoftwareES',
  elasticIndex:
    process.env.NODE_ENV === 'test'
      ? 'softwares_test'
      : process.env.ELASTIC_INDEX_ALIAS,
  elasticType: 'software',
  elasticMapping: mapping,
  elasticClient: new elasticsearch.Client({
    host: `${process.env.ELASTIC_HOST}`,
    apiVersion: '7.3',
    log: 'trace'
  }),
  // elastic mapping does not contain information about is fields are arrays or not
  // so provide this information explicitly for obtaining correct types in GraphQL
  pluralFields: ['categories', 'uploads', 'logo', 'tags']
})

export default {
  softwareES: SoftwareElasticTC.getResolver('search').getFieldConfig(),
  softwarePagination: SoftwareElasticTC.getResolver(
    'searchPagination'
  ).getFieldConfig(),
  softwareConnection: SoftwareElasticTC.getResolver(
    'searchConnection'
  ).getFieldConfig()
}
