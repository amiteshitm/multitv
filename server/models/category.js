import { composeWithMongoose } from 'graphql-compose-mongoose'

import { uploadSchema } from 'server/models/upload'

const mongoose = require('mongoose')
const Schema = mongoose.Schema
const uniqueValidator = require('mongoose-unique-validator')

const CategorySchema = new Schema(
  {
    name: { type: 'String', required: true },
    slug: { type: 'String', required: true, unique: true },

    details: { type: 'String' },
    desc: { type: 'String', required: true },
    is_published: { type: 'Boolean', default: true },

    cover: [uploadSchema],
    logo: [uploadSchema],
    children: [Schema.Types.ObjectId],
    path: [
      {
        name: Schema.Types.String,
        slug: Schema.Types.String,
        category_id: Schema.Types.ObjectId,
      },
    ],
    parent_id: { type: 'ObjectId', default: null },
    features: [{ type: Schema.Types.ObjectId, ref: 'Feature' }],
  },
  { timestamps: { createdAt: 'created_at', updatedAt: 'updated_at' } }
)

CategorySchema.pre('save', async function (next) {
  if (this.parent_id != null) {
    const parent = await CategoryModel.findById(this.parent_id)
    const pathElem = {
      name: parent.name,
      slug: parent.slug,
      category_id: parent._id,
    }
    let path = parent.path
    path.push(pathElem)

    this.path = path
  }
  next()
})

CategorySchema.post('save', async doc => {
  // update the parent
  await CategoryModel.findByIdAndUpdate(doc.parent_id, { $push: { children: doc._id } })
})

CategorySchema.plugin(uniqueValidator)

const CategoryModel = mongoose.model('Category', CategorySchema)

const CategoryTC = composeWithMongoose(CategoryModel, {
  fields: {
    remove: ['created_at', 'updated_at'],
  },
})

export { CategoryModel, CategorySchema, CategoryTC }
