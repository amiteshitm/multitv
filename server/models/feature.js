const mongoose = require('mongoose')
const Schema = mongoose.Schema
const uniqueValidator = require('mongoose-unique-validator')

const featureSchema = new Schema({
  name: { type: 'String', required: true },
  slug: { type: 'String', unique: true },
  data_type: { type: 'String', enum: ['String', 'Boolean'] },
  parent_id: { type: 'ObjectId' },
})

featureSchema.pre('save', function (next) {
  this.slug = this.name.replace(/[^A-Za-z0-9]/, '').toLowerCase()

  next()
})

featureSchema.plugin(uniqueValidator)

const FeatureModel = mongoose.model('Feature', featureSchema)

export { FeatureModel, featureSchema }
