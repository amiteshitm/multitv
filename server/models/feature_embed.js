const mongoose = require('mongoose')
const Schema = mongoose.Schema
const uniqueValidator = require('mongoose-unique-validator')

const featureEmbedSchema = new Schema({
  feature: { type: 'ObjectId', required: true, ref: 'Feature' },
  data: { type: Schema.Types.Mixed },
})

featureEmbedSchema.plugin(uniqueValidator)

const FeatureEmbedModel = mongoose.model('FeatureEmbed', featureEmbedSchema)

export { FeatureEmbedModel, featureEmbedSchema }
