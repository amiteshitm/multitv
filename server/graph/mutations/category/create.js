import { CategoryModel } from 'server/models/category'

const categoryCreate = async (obj, params) => {
  const category = new CategoryModel(params.input)

  const newCategory = await category.save()
  if (!newCategory) {
    throw new Error('Error')
  }
  return newCategory
}

export default categoryCreate
