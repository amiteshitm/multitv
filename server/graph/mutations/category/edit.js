import { CategoryModel } from 'server/models/category'

const categoryUpdate = (obj, params) => {
  const { media, logo, ...vars } = params.input

  return CategoryModel.findOneAndUpdate({ slug: params.input.slug }, vars, { new: true }).then(cat => cat)
}

export default categoryUpdate
