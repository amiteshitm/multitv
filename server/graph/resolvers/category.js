import { schemaComposer } from 'graphql-compose'

import { CategoryTC } from 'server/models/category'

function addCategoryResolvers () {
  schemaComposer.Query.addFields({
    categoryOne: CategoryTC.getResolver('findOne'),
    categoryMany: CategoryTC.getResolver('findMany'),
  })
  createCustomResolvers()
  createRelations()
}

function createCustomResolvers () {
  // TODO: Check if features need to be moved into relation once all is done.
  CategoryTC.addFields({
    features: {
      type: '[Feature]',
      resolve: async obj => {
        const f = await obj.populate('features').execPopulate()

        return f.features
      },
      projection: { features: 1 },
    },
  })
}

function createRelations () {
  // Childre
  CategoryTC.addRelation('children', {
    resolver: () => CategoryTC.get('$findByIds'),
    prepareArgs: {
      _ids: source => source.children,
    },
    projection: { children: 1 },
  })
}

export { addCategoryResolvers }
