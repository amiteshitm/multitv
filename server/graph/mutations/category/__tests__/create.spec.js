import { createTestClient } from 'apollo-server-testing'
import elasticsearch from 'elasticsearch'
import gql from 'graphql-tag'

import { constructTestServer } from 'test/apollo'
import { FactoryBot } from 'test/factories'
import { CategoryModel } from 'server/models/category'

require('test/helper')
require('test/elastic')

jest.mock('elasticsearch')

const CREATE_CATEGORY = gql`
  mutation CategoryCreate($record: CreateOneCategoryInput!) {
    categoryCreate(record: $record) {
      record {
        name
        desc
      }
    }
  }
`

beforeAll(() => {
  elasticsearch.mockImplementation(() => ({
    Client: jest.fn(() => {}),
  }))
})

describe('mutation creates category', () => {
  let server, mutate

  beforeEach(() => {
    server = constructTestServer().server
    mutate = createTestClient(server).mutate
  })

  it('creates parent category', async () => {
    const res = await mutate({
      mutation: CREATE_CATEGORY,
      variables: {
        record: {
          name: 'First Category',
          slug: 'first-category',
          details: 'This is the first category',
          desc: 'This is a category description to test',
        },
      },
    })

    expect(res).toMatchSnapshot()
  })

  it('creates child category', async () => {
    let category = await FactoryBot.create('category')

    const res = await mutate({
      mutation: CREATE_CATEGORY,
      variables: {
        record: {
          name: 'First Category',
          slug: 'first-category',
          details: 'This is the first category',
          desc: 'This is a category description to test',
          parent_id: category.id,
        },
      },
    })

    category = await CategoryModel.findById(category._id)

    expect(category.children).toHaveLength(1)

    expect(res).toMatchSnapshot()
  })
})
