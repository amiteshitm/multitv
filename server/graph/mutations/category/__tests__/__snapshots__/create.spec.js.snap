// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mutation creates category creates child category 1`] = `
Object {
  "data": Object {
    "categoryCreate": Object {
      "record": Object {
        "desc": "This is a category description to test",
        "name": "First Category",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`mutation creates category creates parent category 1`] = `
Object {
  "data": Object {
    "categoryCreate": Object {
      "record": Object {
        "desc": "This is a category description to test",
        "name": "First Category",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
